src/emmy/expression/render.cljc

```diff
@@ -661,27 +644,73 @@
                                 'or (fn [[a b]] (str a " || " b))
                                 '/ render-infix-ratio}))]
-    (fn [x & {:keys [symbol-generator parameter-order deterministic?]
-             :or {symbol-generator (make-symbol-generator "_")
-                  parameter-order sort}}]
-      (let [x      (v/freeze x)
-            params (set/difference (x/variables-in x) operators-known)
-            ordered-params (if (fn? parameter-order)
-                             (parameter-order params)
-                             parameter-order)
-            callback (fn [new-expression new-vars]
-                       (doseq [[var val] new-vars]
-                         (print "  var ")
-                         (print (str var " = "))
-                         (print (R val))
-                         (print ";\n"))
-                       (print "  return ")
-                       (print (R new-expression))
-                       (print ";\n}"))
-            opts {:deterministic?   deterministic?
-                  :symbol-generator symbol-generator}]
-        (with-out-str
-          (print "function(")
-          (print (s/join ", " ordered-params))
-          (print ") {\n")
-          (compile/extract-common-subexpressions x callback opts))))))
+    (fn [x opts]
+      (let [x              (v/freeze x)
+            params         (set/difference (x/variables-in x) operators-known)
+            callback       (fn [new-expression new-vars]
+                             {:params params
+                              :vars   (into [] (map (fn [[k v]] [k (R v)])) new-vars)
+                              :value  (R new-expression)})
+            opts           (merge {:symbol-generator (make-symbol-generator "_")} opts)]
+        (extract-common-subexpressions x callback opts)))))
+
+(defn ->JavaScript
+  "Convert the given expression to a string representation of a
+  JavaScript function.
+
+  Parameters to the function will be extracted from the symbols in the expression.
+  Common subexpression elimination will be performed and auxiliary variables
+  will be bound in the body of the function; the names of these symbols are
+  `_0001, ...`.
+
+  If `:parameter-order` is specified, it is used to determine function parameter
+  order in one of two ways:
+
+  If it is set to a function, that function will be called on the sequence of
+  parameters and is expected to return the parameters in the desired sequence.
+
+  Otherwise, it is interpreted as the sequence of parameters itself. If not
+  specified, the default behavior is `sort`.
+
+  If `:deterministic? true` is supplied, the function will assign variables by
+  sorting the string representations of each term before assignment. Otherwise,
+  the nondeterministic order of hash maps inside this function won't guarantee a
+  consistent variable naming convention in the returned function. For tests, set
+  `:deterministic? true`."
+  [x & {:keys [parameter-order]
```

Comment:
Okay, I remember why I moved this function over. It seemed to me that this function was doing something sympy-like by taking an expression and trying to treat it as a function of the unspecified arguments.

In a similar way to the compile functions, it felt like it might be more consistent to have ->JavaScript take a function, and then generate the arguments and pass them into the fn instead of trying to infer and pull them out (which runs into argument order trouble, if you want to do something other than manually inspect the source before compiling.)

Reply:
We found the same thing. This is why I separated the functions via the map return value: nobody in compile-land looks at the inferred argument list, which is only used by the "printing" version, which I now agree is bogus (or at least that you have to manually deal with the function wrapping).

One way out of this would be to just have the ("legacy") `->JavaScript` just print the function body and leave the construction of the execution context up to the consumer (as compile.cljs does) instead of printing something that probably won't work